src/Processors/QueryPlan/ReadFromMergeTree.cpp

```diff
@@ -658,7 +658,29 @@ Pipe ReadFromMergeTree::spreadMarkRangesAmongStreams(RangesInDataParts && parts_
     {
         /// Reduce the number of num_streams if the data is small.
         if (info.sum_marks < num_streams * info.min_marks_for_concurrent_read && parts_with_ranges.size() < num_streams)
-            num_streams = std::max((info.sum_marks + info.min_marks_for_concurrent_read - 1) / info.min_marks_for_concurrent_read, parts_with_ranges.size());
```

Comment:
couple of thoughts:
* let's not do it if `checkAllPartsOnRemoteFS(parts) == true`
* let's raise the minimum number of marks to 8. as far as I see in your experiment it provided 146.7% vs the maximum of 154% which is almost the same. but it will look less random, because 8 granules is the default block size